core/src/main/scala/kafka/server/ForwardingManager.scala

```diff
@@ -0,0 +1,85 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package kafka.server
+
+import kafka.metrics.KafkaMetricsGroup
+import kafka.network.RequestChannel
+import org.apache.kafka.clients.ClientResponse
+import org.apache.kafka.common.metrics.Metrics
+import org.apache.kafka.common.protocol.Errors
+import org.apache.kafka.common.requests.{AbstractRequest, AbstractResponse, EnvelopeRequest, EnvelopeResponse}
+import org.apache.kafka.common.utils.Time
+
+import scala.compat.java8.OptionConverters._
+
+class ForwardingManager(metadataCache: kafka.server.MetadataCache,
+                        time: Time,
+                        metrics: Metrics,
+                        config: KafkaConfig,
+                        threadNamePrefix: Option[String] = None) extends
+  BrokerToControllerChannelManagerImpl(metadataCache, time, metrics,
```

Comment:
Should we extend the BrokerToControllerChannel here? It looks like we only ever call `forwardRequest`, so maybe it's better for ForwardingManager _not_ to extend BrokerToControllerChannel and instead take it as a dependency. Otherwise we are exposing the `sendRequest` method which we don't really want anyone to use on the `forwardingManager` instance.

Reply:
Actually we do need `sendRequest` on the forwarding manager interface in the topic creation routing (WIP [here.](https://github.com/apache/kafka/pull/9579/). The `forwardRequest` is serving the purpose of embedding client request inside an Envelope, while for certain cases broker would like to send direct CreateTopicsRequest on behalf of its own principal instead of the client principal. I think we could debate whether to rename `forwardingManager` to a different name to also include the possibility of sending direct request, just not have yet come up a better name here.

Reply:
Ah, ok 👍 Thanks for the explanation. Another option would be to let KafkaApis have BrokerToControllerChannelManager and ForwardingManager as separate dependencies. However, we already have lots of dependencies there, so I could see us wanting to use one instance for forwarding and regular requests. I don't really have a strong opinion either way, so let's just leave it as is.